source/game/system/KPadController.cc

```diff
@@ -110,56 +105,53 @@ void RaceInputState::reset() {
     trickRaw = 0;
 }
 
+bool RaceInputState::accelerate() const {
+    return !!(buttons & 0x1);
+}
+
+bool RaceInputState::trickUp() const {
+    return !!(trick & 0x10);
+}
+
 KPadGhostButtonsStream::KPadGhostButtonsStream()
-    : buffer(nullptr), sequenceCount(-1), size(RKG_INPUT_DATA_SIZE), state(2) {}
+    : currentSequence(std::numeric_limits<u32>::max()), state(2) {}
 
 KPadGhostButtonsStream::~KPadGhostButtonsStream() = default;
 
-bool KPadGhostButtonsStream::readIsNewSequence(const u16 *sequence) const {
-    return readSequenceFrames >= (*sequence >> 0x8);
+bool KPadGhostButtonsStream::readIsNewSequence() const {
+    return readSequenceFrames >= (currentSequence & 0xFF);
 }
 
-u16 KPadGhostButtonsStream::readVal(const u16 *sequence) const {
-    return *sequence;
+u8 KPadGhostButtonsStream::readVal() const {
+    return currentSequence >> 8;
 }
 
-u16 KPadGhostButtonsStream::readFrame() {
-    constexpr u32 SEQUENCE_SIZE = 0x2;
-
+u8 KPadGhostButtonsStream::readFrame() {
     if (state != 1) {
         return 0;
     }
 
-    bool bVar1 = false;
-    const u16 *sequence = nullptr;
-
-    if (sequenceCount >= 0) {
-        sequence = buffer + sequenceCount;
-
-        if (readIsNewSequence(sequence)) {
-            bVar1 = true;
-            if (size <= ++sequenceCount * SEQUENCE_SIZE) {
+    if (currentSequence == std::numeric_limits<u32>::max()) {
+        readSequenceFrames = 0;
+        currentSequence = buffer.read_u16();
+    } else {
+        if (readIsNewSequence()) {
+            readSequenceFrames = 0;
+            if (buffer.eof()) {
                 state = 2;
+            } else {
```

Comment:
```suggestion
            if (buffer.eof()) {
                state = 2;
                return 0;
            } else {
```

This eliminates the state check at the end of this block.